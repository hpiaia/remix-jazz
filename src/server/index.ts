export * from './auth'
export * from './request-validator'
export * from './responses'
