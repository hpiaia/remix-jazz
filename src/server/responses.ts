import { json } from '@remix-run/node'

/**
 * Create a response with the status code 201.
 */
export function created<Data = unknown>(data: Data, init?: Omit<ResponseInit, 'status'>) {
  return json(data, { ...init, status: 201 })
}

/**
 * Create a response with the status code 400.
 */
export function badRequest<Data = unknown>(data: Data, init?: Omit<ResponseInit, 'status'>) {
  return json(data, { ...init, status: 400 })
}

/**
 * Create a response with the status code 401.
 */
export function unauthorized<Data = unknown>(data: Data, init?: Omit<ResponseInit, 'status'>) {
  return json(data, { ...init, status: 401 })
}

/**
 * Create a response with the status code 403.
 */
export function forbidden<Data = unknown>(data: Data, init?: Omit<ResponseInit, 'status'>) {
  return json(data, { ...init, status: 403 })
}

/**
 * Create a response with the status code 404.
 */
export function notFound<Data = unknown>(data: Data, init?: Omit<ResponseInit, 'status'>) {
  return json(data, { ...init, status: 404 })
}

/**
 * Create a response with the status code 422.
 */
export function unprocessableEntity<Data = unknown>(data: Data, init?: Omit<ResponseInit, 'status'>) {
  return json(data, { ...init, status: 422 })
}

/**
 * Create a response with the status code 500.
 */
export function internalServerError<Data = unknown>(data: Data, init?: Omit<ResponseInit, 'status'>) {
  return json(data, { ...init, status: 500 })
}
