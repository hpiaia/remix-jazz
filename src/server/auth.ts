import { createCookieSessionStorage, redirect } from '@remix-run/node'
import { unauthorized } from './responses'

type Config = {
  sessionSecret: string
  userSessionKey: string
}

export function createAuthStorage({ sessionSecret, userSessionKey }: Config) {
  const sessionStorage = createCookieSessionStorage({
    cookie: {
      name: '__session',
      httpOnly: true,
      path: '/',
      sameSite: 'lax',
      secrets: [sessionSecret],
      secure: process.env.NODE_ENV === 'production',
    },
  })

  async function getSession(request: Request) {
    const cookie = request.headers.get('Cookie')
    return sessionStorage.getSession(cookie)
  }

  async function signIn({
    request,
    userId,
    expiration,
    redirectTo,
  }: {
    request: Request
    userId: string
    expiration?: number | undefined
    redirectTo: string
  }) {
    const session = await getSession(request)

    session.set(userSessionKey, userId)

    return redirect(redirectTo, {
      headers: {
        'Set-Cookie': await sessionStorage.commitSession(session, { maxAge: expiration }),
      },
    })
  }

  async function signOut({ request, redirectTo }: { request: Request; redirectTo: string }) {
    const session = await getSession(request)

    return redirect(redirectTo, {
      headers: {
        'Set-Cookie': await sessionStorage.destroySession(session),
      },
    })
  }

  async function getUserId(request: Request): Promise<string | undefined> {
    const session = await getSession(request)
    return session.get(userSessionKey)
  }

  async function getUserIdOrThrow(request: Request) {
    const userId = await getUserId(request)
    if (!userId) throw unauthorized('Unauthorized')
    return userId
  }

  return {
    getSession,
    signIn,
    signOut,
    getUserId,
    getUserIdOrThrow,
  }
}
