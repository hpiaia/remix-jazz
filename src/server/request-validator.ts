import { ZodError, type ZodSchema } from 'zod'
import { forbidden, unprocessableEntity } from './responses'

type Options<Data> = {
  schema: ZodSchema<Data>
  authorize?: (request: Request) => boolean | PromiseLike<boolean>
  forbiddenMessage?: string
}

async function getFormData(request: Request) {
  return Object.fromEntries((await request.formData()).entries())
}

export function createRequestValidator<Data>({ authorize, schema, forbiddenMessage }: Options<Data>) {
  async function checkAuthorization(request: Request) {
    const isAuthorized = authorize ? await authorize(request) : true
    if (!isAuthorized) {
      throw await forbidden(forbiddenMessage ?? 'Forbidden')
    }
  }

  return (request: Request) => {
    return {
      async formData() {
        await checkAuthorization(request)

        const parsed = schema.safeParse(await getFormData(request))

        return parsed.success
          ? ({ success: true, data: parsed.data } as const)
          : ({ success: false, errors: parsed.error.flatten() } as const)
      },

      async formDataOrThrow() {
        await checkAuthorization(request)

        try {
          return schema.parse(await getFormData(request))
        } catch (exception) {
          if (exception instanceof ZodError) {
            throw unprocessableEntity(exception.errors)
          }
          throw exception
        }
      },
    }
  }
}
