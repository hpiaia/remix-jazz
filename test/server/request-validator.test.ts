import { z } from 'zod'

import { createRequestValidator } from '../../src/server'
import { createRequest } from '../utils'

describe('request handler', () => {
  const myRequest = createRequestValidator({
    schema: z.object({
      name: z.string(),
      email: z.string().email(),
      password: z.string().min(8),
    }),

    authorize: async (request) => !request.url.includes('nope'),
  })

  it('should validate formData input', async () => {
    const body = new FormData()

    const request = createRequest({ method: 'post', body })

    const { success, errors, data } = await myRequest(request).formData()

    expect(success).toBe(false)
    expect(errors).toBeDefined()
    expect(data).toBeUndefined()

    expect(errors?.fieldErrors).toEqual({
      name: ['Required'],
      email: ['Required'],
      password: ['Required'],
    })
  })

  it('should parse formData input', async () => {
    const body = new FormData()

    body.append('name', 'John Doe')
    body.append('email', 'john@doe.com')
    body.append('password', '12345678')

    const request = createRequest({
      method: 'post',
      body,
    })

    const { success, errors, data } = await myRequest(request).formData()

    expect(success).toBe(true)
    expect(errors).toBeUndefined()
    expect(data).toBeDefined()

    expect(data).toEqual({
      name: 'John Doe',
      email: 'john@doe.com',
      password: '12345678',
    })
  })

  it('should throw unprocessable entity when formData is invalid', async () => {
    const body = new FormData()

    const request = createRequest({
      method: 'post',
      body,
    })

    try {
      await myRequest(request).formDataOrThrow()
    } catch (error) {
      expect(error).toBeInstanceOf(Response)
      expect((error as Response).status).toBe(422)
    }
  })

  it('should throw exception with request body is invalid', async () => {
    const request = new Request('http://localhost', {
      method: 'POST',
      body: 'invalid-body',
    })

    await expect(myRequest(request).formDataOrThrow()).rejects.toThrowErrorMatchingSnapshot()
  })

  it('should throw a 403 response when request is not authorized', async () => {
    const body = new FormData()

    body.append('name', 'John Doe')
    body.append('email', 'john@doe.com')
    body.append('password', '12345678')

    const request = createRequest({
      method: 'post',
      body,
      url: '/nope',
    })

    try {
      await myRequest(request).formDataOrThrow()
    } catch (error) {
      expect(error).toBeInstanceOf(Response)
      expect((error as Response).status).toBe(403)
      expect(await (error as Response).text()).toBe('"Forbidden"')
    }
  })

  it('should pass authorization when no function is present', async () => {
    const noAuthValidator = createRequestValidator({
      schema: z.object({
        name: z.string(),
      }),
    })

    const body = new FormData()
    body.append('name', 'John Doe')

    const request = createRequest({
      method: 'post',
      body,
      url: '/nope',
    })

    const { success, errors, data } = await noAuthValidator(request).formData()

    expect(success).toBe(true)
    expect(errors).toBeUndefined()
    expect(data).toBeDefined()
  })
})
