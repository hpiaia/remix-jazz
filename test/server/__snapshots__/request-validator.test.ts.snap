// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`request handler should throw exception with request body is invalid 1`] = `"Request.formData: Could not parse content as FormData."`;
